### ISSUE TITLE
Issue when compiling with OpenSSL

### ISSUE THREAD

**Issue by hasi** [fpoirotte/phpfarm#16]
Hello,

I installed php 5.6 on Ubuntu 16.04 server and wanted to install php 5.3 as well as I have applications specifically need 5.3 version. Used PHPFarm for that and tried to install based on the tutorial found here http://www.jabommi.de/wiki/downgrade-php-5-5-to-5-3-ubuntu-14-with-multiple-php-versions/.

I couldn't compile PHP and it was stopping at some points and I checked removing extensions one by one from the configuration options file. It was openssl and without that PHP complied fine and working with my applications.

But I got a need to use SMTP mail inside one application and it's not allowing without openssl with PHP5.3. Then I tried to compile PHP again with openssl and got same issues. I tried to use openssl version installed for PHP5.6 version by using the directory path in PHP5.3 configuration options file. But when I provide the path to exact openssl directory it's showing an error about evp.h header file and then I changed the path just to use up to /usr directory based on some solutions found online. But then it's started saying it can't find the OpenSSL libraries. Then I used the below code and managed to get passed that point.

```
--with-openssl=/usr \
--with-libdir=/lib/x86_64-linux-gnu \
```

But now it's stopping at following error.

```
/opt/phpfarm/src/php-5.3.29/ext/openssl/xp_ssl.c:342: undefined reference to 'SSLv3_client_method'
/opt/phpfarm/src/php-5.3.29/ext/openssl/xp_ssl.c:354: undefined reference to 'SSLv3_server_method'
collect2: error: ld returned 1 exit status
Makefile:257: recipe for target 'sapi/cgi/php-cgi' failed
make: *** [sapi/cgi/php-cgi] Error 1
make failed.
```

I found a patch for PHP5.2 version with same error but SSLv2_client_method and SSLv2_server_method. https://bugs.php.net/bug.php?id=54736

Any idea about how to resolve it will be greatly appreciated.

Regards,
Hasi

**Comment by fpoirotte** [comment]
It seems the folks at Ubuntu disabled SSLv3 in recent releases of their openssl package (this is quite logical given that multiple vulnerabilities have been found in the past in SSLv3 and it is now considered insecure).

To be able to compile PHP 5.3 without SSLv3 support, you'll have to apply your own patch.
You can start by making a copying the original "/opt/phpfarm/src/php-5.3.29/ext/openssl/xp_ssl.c" file and repeating the changes (adapted for SSLv3) in the patch attached to the link you posted (https://bugs.php.net/patch-display.php?bug_id=54736&patch=debian_patches_disable_SSLv2_for_openssl_1_0_0.patch&revision=latest).
Once the patch is ready, you can apply it from your "options.sh" script using code like this:
```
pushd "php-$1"
patch -r - -N -p1 < ../custom/icu-pkg-config.diff
popd
```

``$1`` refers to the version string you passed while calling either the compile.sh or main.sh script, eg. "5.3.29-debug". Using pushd/popd makes it possible to apply the patch to several different versions of PHP, if necessary.

**Comment by fpoirotte** [comment]
Thanks for taking the time to file this issue. We believe the problem you reported has been fixed.
It it still persists, feel free to file a new issue.
